substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,814 @@
+	fn is_agent(who: &T::AccountId) -> bool {
+		<Agents<T>>::contains_key(who)
+	}
+
+	/// Returns true if who is delegating to an `Agent` account.
+	fn is_delegator(who: &T::AccountId) -> bool {
+		<Delegators<T>>::contains_key(who)
+	}
+
+	/// Returns true if who is already staking on [`Config::CoreStaking`].
+	fn is_direct_staker(who: &T::AccountId) -> bool {
+		T::CoreStaking::status(who).is_ok()
+	}
+
+	/// Registers a new agent in the system.
+	fn do_register_agent(who: &T::AccountId, reward_account: &T::AccountId) {
+		AgentLedger::<T>::new(reward_account).update(who);
+
+		// Agent does not hold balance of its own but this pallet will provide for this to exist.
+		// This is expected to be a keyless account and not created by any user directly so safe.
+		// TODO: Someday if we allow anyone to be an agent, we should take a deposit for
+		// being a delegator.
+		frame_system::Pallet::<T>::inc_providers(who);
+	}
+
+	/// Migrate existing staker account `who` to an `Agent` account.
+	fn do_migrate_to_agent(who: &T::AccountId, reward_account: &T::AccountId) -> DispatchResult {
+		Self::do_register_agent(who, reward_account);
+
+		// We create a proxy delegator that will keep all the delegation funds until funds are
+		// transferred to actual delegator.
+		let proxy_delegator = Self::sub_account(AccountType::ProxyDelegator, who.clone());
+
+		// Keep proxy delegator alive until all funds are migrated.
+		frame_system::Pallet::<T>::inc_providers(&proxy_delegator);
+
+		// Get current stake
+		let stake = T::CoreStaking::stake(who)?;
+
+		// release funds from core staking.
+		T::CoreStaking::migrate_to_virtual_staker(who);
+
+		// transfer just released staked amount plus any free amount.
+		let amount_to_transfer =
+			T::Currency::reducible_balance(who, Preservation::Expendable, Fortitude::Polite);
+
+		// This should never fail but if it does, it indicates bad state and we abort.
+		T::Currency::transfer(who, &proxy_delegator, amount_to_transfer, Preservation::Expendable)?;
+
+		T::CoreStaking::update_payee(who, reward_account)?;
+		// delegate all transferred funds back to agent.
+		Self::do_delegate(&proxy_delegator, who, amount_to_transfer)?;
+
+		// if the transferred/delegated amount was greater than the stake, mark the extra as
+		// unclaimed withdrawal.
+		let unclaimed_withdraws = amount_to_transfer
+			.checked_sub(&stake.total)
+			.defensive_ok_or(ArithmeticError::Underflow)?;
+
+		if !unclaimed_withdraws.is_zero() {
+			let mut ledger = AgentLedger::<T>::get(who).ok_or(Error::<T>::NotAgent)?;
+			ledger.unclaimed_withdrawals = ledger
+				.unclaimed_withdrawals
+				.checked_add(&unclaimed_withdraws)
+				.defensive_ok_or(ArithmeticError::Overflow)?;
+			ledger.update(who);
+		}
+
+		Ok(())
+	}
+
+	/// Bond `amount` to `agent_acc` in [`Config::CoreStaking`].
+	fn do_bond(agent_acc: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+		let agent = Agent::<T>::get(agent_acc)?;
+
```

Comment:
might be easier to ask `Agent::get` to always to a withdraw unbonded under the hood.

Reply:
Didn't really understand. Could you expand? 
For this pallet, its never really useful to do an implicit withdraw unbonded, but pools do it and it will be handled. Also funds that are withdrawn should be reused for bonding and is tracked in a `unclaimed_withdrawals` register.